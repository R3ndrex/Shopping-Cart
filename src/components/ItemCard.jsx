import { useState } from "react";
import ImageSlider from "./ImageSlider.jsx";
export default function ItemCard({ product, setSelectedItems }) {
    const [inputValue, setinputValue] = useState("");
    function handleSubmit(e) {
        e.preventDefault();
        console.log(1);
        const formData = new FormData(e.currentTarget);
        const item = {
            amount: Number(formData.get(`${product.title}-amount`)),
            title: product.title,
            price: product.price,
        };
        setSelectedItems((prev) => {
            if (prev.some((element) => element.title === item.title)) {
                return prev.map((element) => {
                    if (element.title === item.title)
                        return {
                            ...element,
                            amount: element.amount + item.amount,
                        };
                });
            }
            return [...prev, item];
        });
        setinputValue("");
    }
    return (
        <form onSubmit={handleSubmit} className="store-item">
            <h2>{product.title}</h2>
            <ImageSlider images={product.images} alt={product.title} />
            <input
                placeholder="Amount"
                min={1}
                value={inputValue}
                max={100}
                required
                onChange={(e) => setinputValue(e.currentTarget.value)}
                type="number"
                name={`${product.title}-amount`}
                id={`${product.title}-amount`}
            />
            <button type="submit">Add to cart</button>
        </form>
    );
}
