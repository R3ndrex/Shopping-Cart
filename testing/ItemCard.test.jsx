import { describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import ItemCard from "../src/components/ItemCard";
import userEvent from "@testing-library/user-event";
const product = {
    title: "Title",
    price: 3.33,
    images: ["", "", ""],
};

describe("Item Card", () => {
    it("add correct amount of items", async () => {
        const setSelectedItems = vi.fn();
        const user = userEvent.setup();
        render(
            <ItemCard product={product} setSelectedItems={setSelectedItems} />
        );

        await user.type(screen.getByPlaceholderText("Amount"), "3");
        await user.click(screen.getByText("Add to cart"));

        expect(setSelectedItems).toBeCalledTimes(1);
    });

    it("doesnt add if empty input", async () => {
        const setSelectedItems = vi.fn();
        const user = userEvent.setup();
        render(
            <ItemCard product={product} setSelectedItems={setSelectedItems} />
        );

        await user.click(screen.getByText("Add to cart"));

        expect(setSelectedItems).not.toBeCalled();
    });
    it("changes to default amount after adding to cart", async () => {
        const setSelectedItems = vi.fn();
        const user = userEvent.setup();
        render(
            <ItemCard product={product} setSelectedItems={setSelectedItems} />
        );

        await user.type(screen.getByPlaceholderText("Amount"), "3");
        await user.click(screen.getByText("Add to cart"));

        expect(screen.getByPlaceholderText("Amount")).toHaveTextContent("");
    });

    it("renders correctly", () => {
        const setSelectedItems = vi.fn();
        const { container } = render(
            <ItemCard product={product} setSelectedItems={setSelectedItems} />
        );

        expect(container).toMatchSnapshot();
    });
});
