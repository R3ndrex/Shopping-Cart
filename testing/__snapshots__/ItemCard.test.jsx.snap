// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item Card > renders correctly 1`] = `
<div>
  <form
    class="store-item"
  >
    <h2>
      Title
    </h2>
    <div
      class="relative "
    >
      <svg
        aria-hidden="true"
        class="absolute left-0 h-[3rem] top-1/2"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M7.72 12.53a.75.75 0 0 1 0-1.06l7.5-7.5a.75.75 0 1 1 1.06 1.06L9.31 12l6.97 6.97a.75.75 0 1 1-1.06 1.06l-7.5-7.5Z"
          fill-rule="evenodd"
        />
      </svg>
      <img
        alt="Title"
        class="w-max"
      />
      <svg
        aria-hidden="true"
        class="absolute right-0 h-[3rem] top-1/2"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M16.28 11.47a.75.75 0 0 1 0 1.06l-7.5 7.5a.75.75 0 0 1-1.06-1.06L14.69 12 7.72 5.03a.75.75 0 0 1 1.06-1.06l7.5 7.5Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <input
      id="Title-amount"
      max="100"
      min="1"
      name="Title-amount"
      placeholder="Amount"
      type="number"
    />
    <button
      type="submit"
    >
      Add to cart
    </button>
  </form>
</div>
`;
